crates/bevy_ui/src/render/mod.rs

```diff
@@ -642,7 +655,41 @@ pub fn prepare_uinodes(
         image.id() != DEFAULT_IMAGE_HANDLE.id()
     }
 
-    for extracted_uinode in extracted_uinodes.uinodes.drain(..) {
+    let mut drains: Vec<_> = extracted_uinodes
+        .uinodes
+        .iter_mut()
+        .map(|uinodes| {
+            let mut d = uinodes.drain(..);
+            let first = d.next();
+            (first, d)
+        })
+        .collect();
+
+    let mut next_extracted_node = move || {
+        let mut min_stack_index = usize::MAX;
+        let mut n = usize::MAX;
+
+        for (i, (node, _)) in drains.iter_mut().enumerate() {
+            if let Some(node) = node {
+                if node.stack_index < min_stack_index {
+                    n = i;
+                    min_stack_index = node.stack_index;
+                }
+            }
+        }
+
+        if n == usize::MAX {
+            None
+        } else {
+            let (current, drain) = &mut drains[n];
+            let next = drain.next();
+            let out = current.take();
+            *current = next;
+            out
+        }
```

Comment:
I was at first extremely confused by how this works. So here is a summary for other reviewers:

1. find the index (named `n`) in `drains` where `node.stack_index` is smallest (the `for` loop)
2. Pick the `(current, drain): (Option<T>, Drain<T>)` from `drains`
3. Read the `current`
4. replace it with the `drain.next()`
5. return `current`

I might suggest some improvement in a couple hours.

Reply:
Yes it's weird I just typed this in and it worked immediately, and then all the better-thought-out refactors I tried ruined the performance.